---
title: "Generating fold changes"
output:
  html_document
---

# preliminaries

```{r}
source('load_info.R')
source('../../geuvadis/R/benchmark_methods.R')
source('../../simulation_core/R/simulate_de.R')
ls()
```

Load differential expression results from both sleuth and DESeq2.
Use the genes that were called differentially expressed at FDR 0.05.

## sleuth

Since the likelihood ratio test is more conservative, we will use it only.

```{r}
suppressPackageStartupMessages(library('sleuth'))
sgr <- readRDS('../results/sgr.rds')
sr <- dplyr::filter(sgr$sleuth.lrt, qval <= 0.05)
sr <- dplyr::select(sr, ens_gene)
```

## DESeq2

```{r}
ds <- readRDS('../results/DESeq2.rds')
ds_res <- dplyr::filter(ds, qval <= 0.05)
ds_res <- dplyr::select(ds_res, ens_gene)
```

# merging the results

We can merge them in several different ways.
We decide to merge by the union, but here are some high level statistics:

```{r}
de_summary <- function(res) {
  res <- dplyr::inner_join(res, transcript_gene_mapping, by = 'ens_gene')
  print('The total number of transcripts: ')
  print(nrow(res))
  res <- dplyr::group_by(res, ens_gene)
  res <- dplyr::summarize(res, n_transcripts = length(target_id))
  print('The total number of genes: ')
  print(nrow(res))
  print('The distribution of transcripts per gene: ')
  print(table(res$n_transcripts))
  print(summary(res$n_transcripts))

  invisible(NULL)
}
```

for sleuth:

```{r}
de_summary(sr)
```

for DESeq2:

```{r}
de_summary(ds_res)
```

Merging by the union:

```{r}
all_results_union <- data.frame(ens_gene = union(sr$ens_gene, ds_res$ens_gene),
  stringsAsFactors = FALSE)
de_summary(all_results_union)
```

merging by the intersection:

```{r}
all_results_intersect <- data.frame(
  ens_gene = intersect(sr$ens_gene, ds_res$ens_gene),
  stringsAsFactors = FALSE)
de_summary(all_results_intersect)
```

```{r}
all_results <- all_results_intersect
significant_genes <- dplyr::inner_join(
  transcript_gene_mapping,
  all_results,
  by = 'ens_gene')
```

# getting the fold changes

## getting the ranks within the data set

this is so we can match the ranks later. we don't want something with incredibly
high counts to get a very large fold change from something that had very low counts

```{r}
counts <- kallisto_table(sgr$so)
count_ranks <- dplyr::group_by(counts, target_id)
count_ranks <- dplyr::summarize(count_ranks, est_counts = mean(est_counts))

count_ranks <- dplyr::inner_join(count_ranks, transcript_gene_mapping,
  by = 'target_id')
count_ranks <- dplyr::group_by(count_ranks, ens_gene)
count_ranks <- dplyr::mutate(count_ranks, rank = rank(est_counts, ties = 'first'))
count_ranks <- dplyr::ungroup(count_ranks)
count_ranks <- dplyr::select(count_ranks, -ens_gene)
```

## computing the fold changes

```{r}
fc <- dplyr::summarize(group_by(counts, target_id, condition),
  average_counts = mean(est_counts))
fc <- dplyr::summarize(fc,
  fold_change =
    average_counts[condition == 'B'] / average_counts[condition == 'A'])
fc <- dplyr::inner_join(fc, significant_genes, by = 'target_id')
```

## cleaning up some outliers

note that in R
- 0 / 0 == NaN
- 3 / 0 == Inf

```{r}
fc <- dplyr::mutate(fc, fold_change = ifelse(is.nan(fold_change),
  1, fold_change))
```

If it is infinite (or larger than 5), give it a relatively large fold_change

```{r}
large_fold_change <- quantile(fc$fold_change[is.finite(fc$fold_change)],
  probs = 0.75)
n_infinite <- sum(is.infinite(fc$fold_change))
infinite_replacement <- abs(truncated_normal(n_infinite, large_fold_change))

fc <- dplyr::mutate(fc, is_infinite = is.infinite(fold_change))

fc <- dplyr::mutate(fc, fold_change = ifelse(is.infinite(fold_change),
  infinite_replacement, fold_change))

n_large <- sum(fc$fold_change > 5)
large_replacement <- abs(truncated_normal(n_large, large_fold_change))

fc <- dplyr::mutate(fc, fold_change = ifelse(fold_change > 5, large_replacement,
  fold_change))
```

some merging of data

```{r}
fc <- dplyr::inner_join(fc,
  count_ranks, by = 'target_id')

fc <- dplyr::mutate(group_by(fc, ens_gene),
  n_transcripts = length(ens_gene))
```

dealing with fold changes that are 0 or arbitrarily small

```{r}
small_fold_change <- quantile(fc$fold_change[fc$fold_change > 0], probs = 0.05)
fc <- dplyr::mutate(fc, fold_change = pmax(fold_change, 0.001))
```

compute a natural log transformation

```{r}
fc <- dplyr::mutate(fc, log_fc = log(fold_change))
```

# wrapping up

save everything

```{r}
saveRDS(fc, '../results/fc.rds')
```

## some summaries

```{r}
fc %>%
  ungroup() %>%
  group_by(n_transcripts) %>%
  do({
    q <- quantile(.$fold_change)
    data.frame(quantile = names(q), fold_change = q)
  }) %>%
  print(n = nrow(.))
```

# session info

```{r}
sessionInfo()
```
